### ISSUE TITLE
Per-file shell override?

### ISSUE THREAD

**Issue by ljharb** [vscode-shellcheck/vscode-shellcheck#214]
I have a file that works in multiple shells, and is sourced not executed, so I don't want to put a shebang directive in it (also, a "shell" directive would imply it only works on one shell, and I'm not aware of how i can tell shellcheck i want to check multiple shells). I usually lint it on the command line with `shellcheck -s sh path/to/file && shellcheck -s dash path/to/file && shellcheck -s bash path/to/file` etc.

It would be great if:
1. i could provide a filename in the extension settings, and override the shell (ie, the `-s` value) for it
1. i could provide *multiples* and have the extension warn me about the union of all the warnings

**Comment by felipecrs** [comment]
This is a very interesting use case. I also keep my `.profile` cross-shell (to be sourced by bash and zsh), and what I do is to write a shebang of `#!/bin/sh`, because POSIX is almost always compatible with both.

But you are right about not having a shebang though. Did you consider asking for such feature in the main shellcheck repository? It would be awesome to have a feature like, and it would be more transparent as it would work with all the other editor extensions:

```sh
# shellcheck shell=bash,dash

echo hi
```

**Comment by ljharb** [comment]
I have, in fact - https://github.com/koalaman/shellcheck/issues/664 no progress in 5 years :-(

**Comment by felipecrs** [comment]
@ljharb

Even though your feature request is nice, I still believe that this is a feature to be implemented in the `shellcheck` itself.

Therefore, I'm closing this issue in favor of https://github.com/koalaman/shellcheck/issues/664.

Feel free to comment if you disagree.

**Comment by ljharb** [comment]
I totally agree that would be ideal, but given that that one’s 6 years old, it doesn’t seem likely to be implemented - in which case it’d be pretty helpful to have it here.

**Comment by felipecrs** [comment]
My concern is about the cleanability of the code. I have the impression that such a new feature would not be simple and could add a burden in terms of maintainability (and [we lack maintainers](https://github.com/vscode-shellcheck/vscode-shellcheck/issues/93)).

But I would be happy to be wrong and review any PRs. 😉

Therefore, I will keep it open for a while.

**Comment by diminutivesloop** [comment]
This would be useful for tools like https://github.com/typicode/husky that use shell scripts but ignore shebangs.

**Comment by felipecrs** [comment]
@diminutivesloop, how would it be useful? IIRC husky will always execute the script with `sh` anyway, meaning:

```bash
# shellcheck shell=sh
```

Should be enough, no?

**Comment by diminutivesloop** [comment]
@felipecrs In a shared project I might not have full control over the contents of the husky scripts and other devs who don't use shell-check might not include a shebang. That said it's more of a nice-to-have.

**Comment by felipecrs** [comment]
Ah, ok. That makes total sense. Let me break this down:

1. Allow to set the shell per-file in the extension settings
2. Allow to set multiple shells per-file in the extension settings, and display the aggregated diagnostics of them all

I will fork 2 into a separate issue.
